### PR TITLE
Enabled 'add_cuda' for bool and fixed alpha scalar bug

### LINKED ISSUE #22431: add_cuda not implemented for bool tensor
a = torch.rand(10, 10, device='cuda').bool()
b = torch.rand(10, 10, device='cuda').bool()
print(a - b)

Returns an error: **RuntimeError: "add_cuda" not implemented for 'Bool'**

### PR DESCRIPTION
**Pull request by izdeby**
Enabled 'add_cuda' for bool 
Tested via unit tests

Fixed #22431 #22430

### CODE REVIEW

File: aten/src/ATen/native/cuda/BinaryOpsKernel.cu [hunk 1]
```diff
@@ -13,12 +13,18 @@
 namespace at { namespace native {
 
 void add_kernel_cuda(TensorIterator& iter, Scalar alpha_scalar) {
```
Comment (colesbury) [inline aten/src/ATen/native/cuda/BinaryOpsKernel.cu]: We need a strategy for dealing with "alpha_scalar" with bools. (For CPU too). Require that it's 1?

**Review by colesbury — APPROVED** [review]

File: test/test_torch.py [hunk 2]
```diff
@@ -1695,8 +1695,7 @@ def test_add(self):
         m1 = torch.tensor([True, False, False, True, False, False], dtype=torch.bool)
         m2 = torch.tensor([True, True, False, False, False, True], dtype=torch.bool)
         self.assertEqual(m1 + m2, torch.tensor([True, True, False, True, False, True], dtype=torch.bool))
-
-        # [res] torch.add([res,] tensor1, value, tensor2)
+        self.assertEqual(m1.to('cuda') + m2.to('cuda'), torch.tensor([True, True, False, True, False, True], device='cuda', dtype=torch.bool))
```
Comment (colesbury) [inline test/test_torch.py]: This isn't going to work on non-CUDA builds. Either guard it with a torch.cuda.is_available() or take in a cast fn like some of the other tests do

**Review by colesbury — CHANGES_REQUESTED** [review]
Sorry -- didn't look closely enough. The test_add test needs to be updated. test_neg and / _test_neg are a good example of one way to do this:

https://github.com/pytorch/pytorch/blob/425d28c30abcac22d9bdd2d01cc83e3e0470eb11/test/test_torch.py#L1720-L1730

https://github.com/pytorch/pytorch/blob/425d28c30abcac22d9bdd2d01cc83e3e0470eb11/test/test_torch.py#L1749-L1750

https://github.com/pytorch/pytorch/blob/425d28c30abcac22d9bdd2d01cc83e3e0470eb11/test/test_cuda.py#L1088-L1089

File: test/test_torch.py (hunk 2 above)
Comment (colesbury) [reply test/test_torch.py]: Or maybe `for device in torch.testing.get_all_device_types():` is the preferred way now? I'm not sure. @gchanan would know.

Comment (izdeby) [reply test/test_torch.py]: Yeah, i was about to suggest. I introduced orch.testing.get_all_device_types() as its a bit confusing to test same thing from 2 different test files. Also, test_cuda is really confusing and im planning to take a deep dive there and simplify the whole testing thing

**Comment by ezyang** [comment]
deferring to Sam

File: aten/src/ATen/native/cuda/BinaryOpsKernel.cu [hunk 3]
```diff
@@ -13,12 +13,19 @@
...
-  AT_DISPATCH_ALL_TYPES_AND(kHalf, iter.dtype(), "add_cuda", [&]() {
-    auto alpha = alpha_scalar.to<scalar_t>();
-    gpu_kernel_with_scalars(iter, [alpha]GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
-      return a + alpha * b;
+  if (iter.dtype() == ScalarType::Bool) {
```
Comment (colesbury) [inline aten/src/ATen/native/cuda/BinaryOpsKernel.cu]: This looks the same as the non-bool case. Maybe combine them with AT_DISPATCH_ALL_TYPES_AND2?

**Review by colesbury — APPROVED** [review]

File: aten/src/ATen/native/cuda/BinaryOpsKernel.cu (hunk 1 above)
Comment (xuhdev) [reply aten/src/ATen/native/cuda/BinaryOpsKernel.cu]: Can we enforce such a requirement? This will prevent `-` be interpreted as `||` for bool.

**Pull request closed without merging**